dessia_common/datatools/learning_models.py

```diff
@@ -0,0 +1,1779 @@
+        :type outputs: List[List[float]]
+
+        :param fit_intercept:
+            Whether to fit the intercept for this model. If set to False, no intercept will be used in calculations
+            (i.e. X and Y are expected to be centered).
+        :type fit_intercept: bool, `optional`, defaults to True
+
+        :param positive:
+            When set to True, forces the coefficients to be positive. This option is only supported for dense arrays.
+        :type positive: bool, `optional`, defaults to False
+
+        :param name:
+            Name of `LinearRegression` model
+        :type name: str, `optional`, defaults to `''`
+
+        :return: The Linear model fit on inputs and outputs.
+        :rtype: `LinearRegression`
+        """
+        return cls.fit_(inputs, outputs, name=name, fit_intercept=fit_intercept, positive=positive)
+
+    @classmethod
+    def fit_predict(cls, inputs: Matrix, outputs: Matrix, predicted_inputs: Matrix, fit_intercept: bool = True,
+                    positive: bool = False, name: str = '') -> Tuple['LinearRegression', Union[Vector, Matrix]]:
+        """ Fit outputs to inputs and predict outputs for `predicted_inputs`: succession of fit and predict. """
+        return cls.fit_predict_(inputs, outputs, predicted_inputs, name=name,
+                                fit_intercept=fit_intercept, positive=positive)
+
+
+class Tree(Model):
+    """
+    Base object for handling a scikit-learn tree._tree.Tree object (Cython).
+
+    Please refer to https://scikit-learn.org/stable/auto_examples/tree/plot_unveil_tree_structure.html for more
+    information on attributes of Tree object and understanding the decision tree structure for basic usage.
+
+    :param n_classes:
+        Number of output classes to predict from data.
+    :type n_classes: List[int], `optional`, defaults to `None`
+
+    :param n_features:
+        Number of features to handle in data.
+    :type n_features: int, `optional`, defaults to `None`
+
+    :param n_outputs:
+        The number of outputs when fit is performed.
+    :type n_outputs: int, `optional`, defaults to `None`
+
+    :param tree_state:
+        All required values to re-instantiate a fully working scikit-learn tree are stored in this parameter.
+    :type tree_state: Dict[str, Any], `optional`, defaults to `None`
+
+    :param name:
+        Name of Tree
+    :type name: str, `optional`, defaults to `''`
+    """
+
+    def __init__(self, n_classes: List[int] = None, n_features: int = None, n_outputs: int = None,
+                 tree_state: Dict[str, Any] = None, parameters: Dict[str, Any] = None, name: str = ''):
+        self.n_classes = n_classes
+        self.n_features = n_features
+        self.n_outputs = n_outputs
+        self.tree_state = tree_state
+        Model.__init__(self, parameters=None, name=name)
+
+    def _data_hash(self):
+        hash_ = npy.linalg.norm(self.tree_state['values'][0])
+        hash_ += sum(self.n_classes)
```

Comment:
Actually, i don't really care. The purpose is to rapidly eliminate non equal trees, because their == is long (because they are list of list of list).

Are you ok with my answer ?